### ISSUE TITLE
PostgreSQL support

### ISSUE THREAD

**Issue by codefromthecrypt** [openzipkin/zipkin#1232]
_From @adriancole on March 11, 2016 1:19_

@Koizumi85 asked on gitter about PostgreSQL support. spanstore-jdbc takes advantage of some SQL syntax which might not be supported in PostgreSQL. Before we start down this path, let's figure out some basics. For example, which version of PostgreSQL is needed, so that we can test the configuration desired.

_Copied from original issue: openzipkin/zipkin-java#93_

**Comment by codefromthecrypt** [comment]
@Koizumi85 nag..

**Comment by codefromthecrypt** [comment]
_From @eirslett on March 21, 2016 10:21_

Could it be supported as an external/optional module? I'm not sure it's a good idea to reuse the JDBC code, even if that's tempting.

**Comment by codefromthecrypt** [comment]
_From @Koizumi85 on March 21, 2016 10:28_

Actually i think with postgresql 9.5 the actual code could possibly work if there were a possibility to switch the database type in Jooq..
Background: psql9.5 supports "upsert" semantics which are heavily used here.
But postgresql 9.5 is fairly new and is not hosted in the repositories of Ubuntu LTS and other linux distributions and there are a lot of sysadmins who don't want to take the postgresql repositories for LTS reasons.

**Comment by codefromthecrypt** [comment]
_From @eirslett on March 21, 2016 12:11_

9.1 or 9.3 are the safest choices, I think.

**Comment by codefromthecrypt** [comment]
_From @Koizumi85 on March 21, 2016 13:10_

9.1 and 9.3 should be almost identically unless the json column type of 9.3. but i don't think we would need this here, or?
So I think a 9.1 compatibility would be nice.

**Comment by codefromthecrypt** [comment]
_From @eirslett on March 21, 2016 13:17_

Sounds good to me. The JSON query capabilites would be nice to have (for things like annotations), but it wasn't added in MySQL until 5.7.8, I think many linux distros are at MySQL 5.5 or so. And the existing SQL span store doesn't use JSON.

**Comment by dwelch2344** [comment]
Any word on this? We dev against PG and would love to not have another database dep just for zipkin (especially on dev machines)

**Comment by codefromthecrypt** [comment]
your are lucky number 3 person to be interested in this! https://blog.codinghorror.com/rule-of-three/

if we started this, would you help maintain and/or test it? Which version of postgresql can you use?

**Comment by dwelch2344** [comment]
Yes! Definitely. Our company is about to launch and one of our goals is to contribute to active open source projects, so we have buy-off from an organizational level to contribute to FOSS. 

We're on 9.5+, but we could easily support earlier versions. I agree to your previous statements that 9.1 would be a good goal to shoot for.

**Comment by codefromthecrypt** [comment]
ooo this sounds suspiciously like an offer to help :) @dwelch2344 would you be up to designing this?

ps the version discussion is dated. ex 9.4 is the current supported version https://www.postgresql.org/download/linux/ubuntu/

@Koizumi85 @eirslett @klette @mansu @jcarres-mdsol any thoughts on this one?

**Comment by hwasungmars** [comment]
Is there any update on this front?  I would like to see PostgreSQL support for Zipkin as well.

**Comment by codefromthecrypt** [comment]
no one is actively working on it quite yet, hoping someone can pick it up. Regardless, it is probably best started as a separate repo I think.

**Comment by codefromthecrypt** [comment]
best as a separate repo as we are pretty out of resources on this one. adding another DB might hurt our memory and bump build execution past 20m. that said, maybe circleci might make that better.. wdyt @abesto

**Comment by codefromthecrypt** [comment]
First, we need to pick a minimum version Ex 9.5 supports upsert and if we use the existing schema from mysql, it would be the easiest to port. On that note, we should decide if we are just porting, or looking to design a high performance schema.

IOTW is the motivation just using postgresql? or are we trying to make a high-performance alternative that uses postgresql?

**Comment by hwasungmars** [comment]
I think most companies don't really need Cassandra on the backend.  I suspect a high-performance alternative for medium-size data will something that will appeal to this audience.  What do you think?

**Comment by hwasungmars** [comment]
I would happily volunteer, but I don't have enough knowledge of Zipkin nor PostgreSQL to take on the task...

**Comment by Koizumi85** [comment]
I agree with @hwasungmars. Implementing PostgreSQL support would be a good opportunity to provide a high-performance implementation backed by a relational database. I suppose there are a lot of people who aren't used to NoSQL databases yet and would appreciate a relational backend which provides a good performance

For the discussion regarding version to support: as on most servers run LTS versions of ubuntu or something like this and there are a lot of distributions only delivering 9.3 with their inbuild-repositories (ubuntu for example until 18.04 LTS if I am not wrong) it would be a good decission I think. But I see, that upsert capabilities of 9.5 would make it a lot easier to implement.

**Comment by codefromthecrypt** [comment]
> I think most companies don't really need Cassandra on the backend. I
> suspect a high-performance alternative for medium-size data will something
> that will appeal to this audience. What do you think?

no insight into deployments of cassandra vs elasticsearch vs postgresql to
be honest

**Comment by eirslett** [comment]
> I think most companies don't really need Cassandra on the backend. I suspect a high-performance alternative for medium-size data will something that will appeal to this audience. What do you think?

I believe Cassandra is the most widely used Zipkin backend for production workloads.

> no one is actively working on it quite yet, hoping someone can pick it up. Regardless, it is probably best started as a separate repo I think.

This is a good idea. Historically, we've already abandoned one DB backend (HBase, the support for that was eventually discontinued), and having implementations in separate repositories means it's up to module maintainers to make sure the backends are up to date. If support for ObscureDB/FadDB is implemented, and then the original authors lose interest and/or nobody else wants to maintain it, then the burden will not be in the core team, it will be on the developers who want to run Zipkin on XxxDB. (I'm not saying PostgreSQL is obscure, it's widely used in the tech world)

There is a balance to strike, between supporting modules in the core repo or separate repos. MySQL was chosen as the relational/simple-to-setup SQL alternative, and Cassandra for the high-scale deployment solution. Elasticsearch was recently added to leverage some of the query functions supported in Lucene (I think).

**Comment by ys** [comment]
I sadly have not much to give but would love to see a postgres backend:P

**Comment by ys** [comment]
I would be happy to try to review schema ideas/implementations. Do not hesitate to ping me if that happens.

**Comment by varyumin** [comment]
I too would like see a PostgreSQL backend

**Comment by tajdars** [comment]
Is there any work going on for postgres persistence support ? Any way I can help ?

**Comment by codefromthecrypt** [comment]
work here needs champions to run it through. If you like, you can look at
the work-on-progress voltdb experiment as an example, though in the schema
likely completing normalization of the data is preferred (unclear on this
really)
https://github.com/adriancole/zipkin-voltdb

Anyway, you can create a repo based on this to start toying with it. Right
now, no one has worked on postgres

**Comment by tajdars** [comment]
Came across this :

https://github.com/IAMTJW/zipkin/tree/master/zipkin-storage/postgresql

Any chance of this getting merged(accepted) ?

**Comment by codefromthecrypt** [comment]
>
> Came across this :
>
> https://github.com/IAMTJW/zipkin/tree/master/zipkin-storage/postgresql
>
> Any chance of this getting merged(accepted) ?
>

first, the author would need to volunteer to contribute the code.
technically though, this looks like a port of the mysql code, which isn't a
good schema to emulate. so, in its form I dont expect that specific code to
be merged.

>
